src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs

```diff
@@ -514,6 +556,111 @@ public class OpenIdConnectAuthenticationHandler : AuthenticationHandler<OpenIdCo
         }
 
         /// <summary>
+        /// Redeems the authorization code for tokens at the token endpoint
+        /// </summary>
+        /// <param name="authorizationCode">The authorization code to redeem.</param>
+        /// <param name="redirectUri">Uri that was passed in the request sent for the authorization code.</param>
+        /// <returns>OpenIdConnect message that has tokens inside it.</returns>
+        protected virtual async Task<OpenIdConnectMessage> RedeemAuthorizationCode(string authorizationCode, string redirectUri)
+        {
+            var openIdMessage = new OpenIdConnectMessage()
+            {
+                ClientId = Options.ClientId,
+                ClientSecret = Options.ClientSecret,
+                Code = authorizationCode,
+                GrantType = "authorization_code",
+                RedirectUri = redirectUri
+            };
+
+            var requestMessage = new HttpRequestMessage(HttpMethod.Post, _configuration.TokenEndpoint);
+            requestMessage.Content = new FormUrlEncodedContent(openIdMessage.Parameters);
+            var responseMessage = await Backchannel.SendAsync(requestMessage);
+            responseMessage.EnsureSuccessStatusCode();
+            var tokenResonse = await responseMessage.Content.ReadAsStringAsync();
+            var jsonTokenResponse = JObject.Parse(tokenResonse);
```

Comment:
They are 2 different extensibility points. RedeemAuthorizationCode would let users control how a code should be redeemed. RunAuthorizationCodeRedeemedNotificationAsync let users control what to do with the tokens, cahing access or refresh tokens if they want. Putting the notification inside the redemption method would make a user overriding RedeemAuthorizationCode ensuring that code redeemed notification is called too. I don't want users to think about notification if all they want to do is override redemption method.